### PR TITLE
Add gf32 module

### PR DESCRIPTION
**Pull request by tcharding**
We are attempting to get this crate to v0.1 (ie, API stabilization).

Currently we use a type call `u5` for the 5 bit math, while valid this is slightly misleading because our `u5` is not basic integer type. Within the bech32 encoding scheme the 5 bit type is an element of GF32 (Galois Field).

Add a `gf32` module that defines a type `Fe32` (Field Element). This type is similar to the current `u5` but includes add/sub/div/mul over the finite field.

Introduce the `primitives` module to hold the `gf32` module (and other future primitive types).

As yet do not make use of the `Fe32` type. Doing so is trivial and will be done later after more primitive types are added.

### Note on attribution

I'd like to be conscious not to make you feel like we are clobbering your crate clarkmoody, as such I left an attribution in this new file to you even though I copied the code from another crate that apoelstra had written. I felt this was correct because if this merges and eventually replaces the `u5` type the code is really similar to what you wrote originally. Also we are keeping your chosen MIT license and re licensing any code we pull from other places (I'm speaking for apoelstra really because the CC license is his favoured license but I believe I understand his wishes on this issue since we have discussed it before).

### CODE REVIEW

**Comment by tcharding** [comment]
Andrew and I have gotten pretty psyched over the last few days and hacked up a bunch of stuff, if anyone is interested to look, check out `04-04-primitives-api` branch on my `rust-bech32` tree. Its WIP, but that is the direction we are hoping to push.

**Comment by apoelstra** [comment]
cc @clarkmoody we won't merge this without your signoff, since it's part of a larger project that substantially rewrites the library in terms of iterator adaptors.

(I edited the PR description to remove all the @ signs.)

File: Cargo.toml
Comment (apoelstra) [inline Cargo.toml:7]: In 310012dd3e5f5725b5bde9021606ed1d5542620d:

I'd say "and implement the bech32 and bech32m checksums" rather than putting these in parens.

File: Cargo.toml
Comment (apoelstra) [inline Cargo.toml:9]: Maybe also add "bitcoin" and "cryptocurrency" here. I forget if there's a limit on the number of keywords you can have.

File: src/primitives/gf32.rs
Comment (apoelstra) [inline src/primitives/gf32.rs:110]: In 310012dd3e5f5725b5bde9021606ed1d5542620d:

lol multiple of these misspell `bech32`

File: src/primitives/gf32.rs
Comment (apoelstra) [inline src/primitives/gf32.rs:66]: In 310012dd3e5f5725b5bde9021606ed1d5542620d:

I'd remove this comment. Every module in `primitives` will probably use the internal `u8`.

File: src/primitives/gf32.rs
Comment (apoelstra) [inline src/primitives/gf32.rs:186]: In 310012dd3e5f5725b5bde9021606ed1d5542620d:

I think we can remove the cast and do `.into`. Since Rust 1.5 we've been able to convert from u8 this way (and since 1.26 from u16).

File: src/primitives/gf32.rs
Comment (apoelstra) [inline src/primitives/gf32.rs:193]: In 310012dd3e5f5725b5bde9021606ed1d5542620d:

Why do we have these methods when we already have operator overloads?

File: src/primitives/gf32.rs
Comment (apoelstra) [inline src/primitives/gf32.rs:227]: In 310012dd3e5f5725b5bde9021606ed1d5542620d:

IMO we should just drop this impl. Converting from a *string* to a single field element is a super weird thing to do, and I don't use this anywhere in my code (because I'd forgotten I"d implemented it :)).

Dropping this would also let us drop the `ExtraChar` and `InvalidLength` variants from the Error enum.

File: src/primitives/gf32.rs
Comment (apoelstra) [inline src/primitives/gf32.rs:280]: I think the negative check is entirely superfluous since `u8::try_from` will do it for us. So we can drop that, and drop the `Error::Negative` variant.

File: src/primitives/gf32.rs
Comment (apoelstra) [reply src/primitives/gf32.rs:397]: Also we should maybe link to the Codex32 booklet https://github.com/BlockstreamResearch/codex32/blob/master/SSS32.ps when citing "the multiplication disk".

Comment (apoelstra) [reply src/primitives/gf32.rs:397]: I guess on a comment at the top of this entire unit test. And on `recovery_wheel`.

**Comment by apoelstra** [comment]
Done reviewing 310012dd3e5f5725b5bde9021606ed1d5542620d. Sorry for having so many nits on code that I largely wrote myself :P but I didn't really look at it after writing it.

File: Cargo.toml
Comment (tcharding) [reply Cargo.toml:9]: Added. I added this too:
`categories = ["encoding", "cryptography::cryptocurrencies"]`

File: src/primitives/gf32.rs
Comment (tcharding) [reply src/primitives/gf32.rs:110]: ha!

**Comment by tcharding** [comment]
What is the reason we do not implement `PartialOrd` and `Ord`? Is this a finite field math thing?

**Comment by tcharding** [comment]
All suggestions implemented.

**Comment by apoelstra** [comment]
> What is the reason we do not implement `PartialOrd` and `Ord`? Is this a finite field math thing?

Yes. There is no meaningful partial ordering on partial field elements that I can think of. Certainly not a natural one that people would use `<` or `>` to refer to.

**Review by apoelstra — APPROVED** [review]
ACK bd010e1f74f8483c8735c5005549894bf51cb7d1

File: src/primitives/gf32.rs
Comment (clarkmoody) [inline src/primitives/gf32.rs:276]: Other than the doc comment, it looks like these implementations are identical.

File: src/primitives/gf32.rs
Comment (clarkmoody) [inline src/primitives/gf32.rs:178]: Why not cast to `usize` here instead of below?

Comment (apoelstra) [reply src/primitives/gf32.rs:178]: Good call. I think it we just iterated too many times and wound up with goofy code. Agreed, casting to `usize` directly would be better.

File: src/primitives/gf32.rs
Comment (apoelstra) [reply src/primitives/gf32.rs:276]: :) you're right -- we used to have an explicit "negative number" check for the signed variants but we realized that `u8::try_from` would do this check for us and removed it.

So we should combine the macros now.

**Comment by apoelstra** [comment]
cc @tcharding can you fix the two nits that Clark highlighted?

**Comment by tcharding** [comment]
Oh damn, I missed your review last week @clarkmoody, sorry man totally unintentional - I'm on it now!

File: src/primitives/gf32.rs
Comment (tcharding) [reply src/primitives/gf32.rs]: That was all me, implementing review suggestions wildly without looking close enough at what I was doing :) Thanks, good reviewing.

File: src/primitives/gf32.rs
Comment (tcharding) [reply src/primitives/gf32.rs]: Changing the cast made me read this code again and think a bit harder. I then simplified it to use explicit value checks instead of using the type conversions to implicitly ensure the array access is within bounds. Next I added a kani verification and fixed the bugs I'd introduced - profit!

**Comment by tcharding** [comment]
Changes in force push:
- Implemented review suggestions
- A few minor docs improvements
- kani verification test

File: src/primitives/gf32.rs
Comment (apoelstra) [inline src/primitives/gf32.rs:179]: In 45b7cd8:

This cast might truncate. Using `usize::try_from` would be better. Or you could use `i8::try_from`, like the original code did, followed by `usize::try_from`, if you want to be fancy.

File: src/primitives/gf32.rs
Comment (apoelstra) [inline src/primitives/gf32.rs:456]: Nice!

File: src/primitives/gf32.rs
Comment (clarkmoody) [reply src/primitives/gf32.rs:179]: Can we trick the test suite into giving us 16-bit `usize`?

**Review by clarkmoody — APPROVED** [review]
ACK 45b7cd89772c41222f6c2164e51a1cbed98df99e

File: src/primitives/gf32.rs
Comment (apoelstra) [reply src/primitives/gf32.rs:179]: That would be cool, but no, not that I'm aware of.

Comment (tcharding) [reply src/primitives/gf32.rs:179]: I thought we didn't support 16 bit machines? Want me to add
```rust
// Disable 16-bit support at least for now as we can't guarantee it yet.
#[cfg(target_pointer_width = "16")]
compile_error!(
    "rust-bitcoin currently only supports architectures with pointers wider than 16 bits, let us
    know if you want 16-bit support. Note that we do NOT guarantee that we will implement it!"
);
```

??

Comment (tcharding) [reply src/primitives/gf32.rs:179]: With `s/rust-bitcoin/rust-bech32/`

File: src/primitives/gf32.rs
Comment (tcharding) [reply src/primitives/gf32.rs:456]: Found bugs too, it was mad!

File: src/primitives/gf32.rs
Comment (apoelstra) [reply src/primitives/gf32.rs:179]: No, I think we should just support 16-bit machines, since it has zero cost for this crate.

Comment (tcharding) [reply src/primitives/gf32.rs:179]: I'm not sure this is much better than the original but I've left it as:
```rust
    /// Creates a field element from a single bech32 character.
    pub fn from_char(c: char) -> Result<Fe32, Error> {
        // i8::try_from gets a value in the range 0..=127 since char is unsigned.
        let byte = i8::try_from(u32::from(c)).map_err(|_| Error::InvalidChar(c))?;
        let ascii = usize::try_from(byte).expect("byte value within range 0..=127");
        // We use -1 for any array element that is an invalid char to trigger error from u8::try_from
        let u5 = u8::try_from(CHARS_INV[ascii]).map_err(|_| Error::InvalidChar(c))?;
        Ok(Fe32(u5))
    }
```

**Review by apoelstra — APPROVED** [review]
ACK 67dab80cd01c7a3c95fd54d7d84b46eff7bd9e98

**Review by apoelstra — APPROVED** [review]
ACK 8a8d33d385838dc8c785e5a48f816ba8a4b72c9d

**Comment by apoelstra** [comment]
@clarkmoody can you do a quick re-ACK? the diff is pretty small between this and the last commit that you did ACK.

**Review by apoelstra — APPROVED** [review]
ACK 9f59765ead4b03fae85ca5499f330a7da8af361a

**Review by clarkmoody — APPROVED** [review]
ACK 9f59765ead4b03fae85ca5499f330a7da8af361a

**Pull request merged**
